### DISCUSSION TITLE
Scrripts gen:types or gen:importmap do not run if type is not module inside package.json

### DISCUSSION THREAD

**Discussion by HQ92** [payloadcms/payload#9807] [Q&A]
### Describe the Bug

I have a use-case where I am running a payload admin panel inside a nextjs project with a babel.config.js as I am using StyleX which does not work with SWC at the moment. Removing type="module" from the package file causes payload commands to fail. But adding it causes nextjs to fail because it cannot import babel.config because internally it uses require to load in custom babel.configs.

Is there a way to get payload commands to run without type='module', or some way to get the project to work with a custom babel.config with type='module'?

### Link to the code that reproduces this issue

Any new payload v3 project, remove, type='module', add simple babel.config.js to root.

### Reproduction Steps

1. Create new starter Payload v3 project.
2. Add custom babel.config.js, with the simplest config.
3. Remove type='module'
4. Try to run gen:types or gen:importmap
5. Will fail

### Which area(s) are affected? (Select all that apply)

area: core

### Environment Info

```text
Payload: v3.2.1
Node: v21.1.0
Nextjs: 15.0.3
```

**Comment by AlessioGr** [comment]
Payload requires `type: module` and at the moment we don't officially support CommonJS and have no plans to do so. I'll move this to a discussion since it's not an "issue" given that CJS is not officially supported by us.

Payload will run fine without type: module within Next.js, as the Next.js Bundler handles ESM/CJS interop for you. But yea, this will not be the case outside of Next.js, e.g. when running custom scripts.

One way to get this to work would be to use a runtime that supports CJS/ESM interop for running the payload scripts. Some options are Bun, Node.js 23 with their new `require(esm)` support, and maybe `tsx`.

However, I would recommend keeping the "type: module" and finding a way for Next.js to load the babel config. I think if you simply rename it to `babel.config.cjs` and use CJS-syntax in your babel config, Next.js should be able to load it just fine. The .cjs file extension should take precedence over "type: module" and force it to be loaded through CommonJS

**Reply by HQ92** [reply]
Thanks for the answer, I've tried to rename babel.config, but renaming it to any cjs or similar makes nextjs panic and it says it can't load it and to change it back, I've tried a whole bunch of ways around this but no luck. Currently I'm resigned to not having type module, and manually regenerating types when needed by adding that line into package, running command, then removing the line. Not ideal, but type gen is likely to be just infrequent enough that it's ok to deal with whilst I hope stylex can move off of babel.

Thanks anyways.

**Reply by ZuBB** [reply]
hey @AlessioGr, I have a Q related to this.

When we do `npx create-payload-app ...` bootstrapping of the project (read `package.json`) is under our control, so we are able to go with `type: "module"` on our own.

But in docs there is a note that `create-payload-app` can be used within existing projects, where usage of `type: "module"` might not be possible. What to do in that case? Does that mean that the user is not able to use any of Payload's built-in "commands" (generate types, import map or migration)?
